### ISSUE TITLE
Stack probe

### ISSUE THREAD

**Issue by antoyo** [Xudong-Huang/may#29]
Hi.
To have better safety, could we use guard pages and [stack probes](https://manishearth.github.io/rust-internals-docs/compiler_builtins/probestack/index.html) in `may`?
I'm not sure it's possible, just asking.
Thanks.

**Comment by Xudong-Huang** [comment]
thanks for the link!
currently `May` doesn't use guard page for coroutines stacks, this need alloc extra one more page for each coroutine. and it has a very unsafe way to probe the stack overflow implemented by checking the footprint of the coroutine stack. So this is why exceed stack could cause undefined behavior.

I check the `__rust_probestack` src code and find that this function needs to be called in the user's code, not in the `May` library, because it only used for one single function frame to detect stack overflow, and for libary design we can't have any information how the stack frames would grow. If uses not call this for a large stack frame (bigger than one page) function it still triggers undefined behavior even we has a stack gurad page.

**Comment by vitiral** [comment]
> I check the __rust_probestack src code and find that this function needs to be called in the user's code, not in the May library, because it only used for one big single function frame to detect stack overflow, and for libary design we don't have any information about how the stack frames would grow.

I may not be understanding, but if it _has_ to be called "in the user's code" then couldn't the `go!` macro call it?

**Comment by Xudong-Huang** [comment]
@vitiral: `go!` macro is just a thin wrapper, which doesn't need stack probe at all. only those leaf functions that have very big stack frame would be benefit from  `__rust_probestack ` when they are running with a `guard page` on the stack. But may doesn't have guard pages for coroutines now, so call `__rust_probestack`  is no help.

**Comment by vitiral** [comment]
I feel the big issue with `may` is that it does not guard against undefined behavior. I think most use cases don't care that they have to limit their stack size (most co-routines will spawn a single function call!) and wouldn't mind if the coroutine _paniced_ in that case. Undefined behavior, however, is a huge problem.

Is it possible to prevent undefined behavior and instead panic... or even abort?

**Comment by Xudong-Huang** [comment]
yes, it's possible to add a stack guard page for each coroutine. And this would need an extra one page memory allocated. for most cases this is not a big issue which I preferred to add the behavior in future. if stack overflow detected, OS will kill the program with a segment fault(maybe with better output by rust, this is not a UB, I think) just like normal thread stack overflow implemented by rust.

**Comment by vitiral** [comment]
cool, thanks for the response!

I think it makes sense for `go!` to use the stack probe, but allow users to access a coroutine without probed memory (through an unsafe function) if they really want to squeeze performance.

Thanks!

**Comment by killme2008** [comment]
I think this is a serious issue that prevent people to use may in production.People are not afraid  of panic but undefined. :smile: 

Do you have any plan to fix this issue?

**Comment by Xudong-Huang** [comment]
Sorry about this delay. I have a solution. But not implement it. You are right, we should have it as soon as posdible🙀

发自我的 iPhone

在 2019年4月23日，10:36，dennis zhuang <notifications@github.com<mailto:notifications@github.com>> 写道：


I think this is a serious issue that prevent people to use may in production.People are not afraid of panic but undefined. 😄

Do you have any plan to fix this issue?

**Comment by Xudong-Huang** [comment]
now the generator lib use guarded stack, may will automatically has the desired behavior.

ref  https://github.com/Xudong-Huang/generator-rs/issues/12
